### PR TITLE
[MRv2][Feature][310P] Adapting MRv2 for Qwen3-VL+Qwen3.5 on the 310P

### PR DESCRIPTION
**Pull request by Thiagor2002**
### What this PR does / why we need it?

This PR adds Model Runner V2 support for Qwen3-VL and Qwen3.5 on Ascend 310P , including these features:

* Tensor Parallelism
* Eager and ACLGraph execution
* NZ-format KV cache
* Triton-free request states, sampling, block-table handling, and input preparation

so that 310P can now support these models via MRv2:
1. Qwen3-Dense/VL/MoE, such as Qwen3-8B, Qwen3-VL-2B/4B/8B, Qwen3-30B-A3B
2. Qwen3.5-Dense/MoE, such as Qwen3.5-2B/4B/9B/27B, Qwen3.5-35B-A3B
3. Qwen3-w8a8/w8a8sc, such as Qwen3-8B-w8a8/w8a8sc, Qwen3-VL-4B/8B-w8a8sc-vllm, Qwen3-30B-A3B-w8a8
4. Qwen3.5-w8a8_dynamic, such as Qwen3.5-2B/4B/9B/27B-W8A8

RFC: #15577

### Does this PR introduce _any_ user-facing change?

Yes. Ascend 310P users can enable Model Runner V2 for supported Qwen3-VL and Qwen3.5 models.

### How was this patch tested?

- vLLM main: https://github.com/vllm-project/vllm/commit/ba07e4a48fc951300d97eb506217dd530583dea3

### CODE REVIEW

File: vllm_ascend/_310p/ops/qwen3vl_310.py [hunk 1]
```diff
@@ -16,6 +16,33 @@
 #
 
 import torch
+from vllm.model_executor.models.qwen3_vl import pos_embed_interpolate_native
+
+
+def fast_pos_embed_interpolate_310(self, grid_thw: list[list[int]]) -> torch.Tensor:
```
Comment (YangShuai52) [inline vllm_ascend/_310p/ops/qwen3vl_310.py]: The 310P will not install Triton; it will automatically select the upstream native implementation of VLLM, so there should be no need to rewrite it here.

File: vllm_ascend/_310p/quantization/methods/w8a8_dynamic.py [hunk 2]
```diff
@@ -143,31 +154,21 @@ def apply(
...
-        if need_unsqz:
-            output = output.unsqueeze(dim=1)
-        return output
+        # Always ND [N, K] + precomputed fp16 dequant. torch.compile/GE cannot
+        # safely run 310P QuantBatchMatmulV3_NZ_NZ on these dynamic-quant linears.
+        bias_term = bias if (tp_rank is None or tp_rank == 0) else None
+        weight_fp = layer.weight_fp
+        if weight_fp.dtype != x.dtype:
+            weight_fp = weight_fp.to(x.dtype)
+        return torch.nn.functional.linear(x, weight_fp, bias_term)
```
Comment (YangShuai52) [inline vllm_ascend/_310p/quantization/methods/w8a8_dynamic.py]: 这里为什么回退到伪量化了，不应该这么做，Qwen3.5的W8A8量化是动态量化

File: vllm_ascend/_310p/worker/v2/attn_utils.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+reshape. 310P V2 allocates NZ tensors in ``NPUModelRunner310V2`` and never
+calls the shared reshape path.
+"""
+
+from __future__ import annotations
+
+from typing import Any
+
+
+def get_310p_non_mla_kv_cache_shapes(
```
Comment (YangShuai52) [inline vllm_ascend/_310p/worker/v2/attn_utils.py]: Looking closely, I don't think this needs to be included, because the 310P doesn't currently support asymmetric-head models.

File: vllm_ascend/_310p/worker/v2/model_runner.py [hunk 4]
```diff
@@ -291,8 +293,22 @@ def finish_requests(self, scheduler_output: SchedulerOutput) -> None:
             # the new layout is gathered and copied to attention metadata.
             torch.npu.current_stream().synchronize()
 
+    def get_kv_cache_spec(self) -> dict[str, KVCacheSpec]:
```
Comment (YangShuai52) [inline vllm_ascend/_310p/worker/v2/model_runner.py:312]: Is everything working correctly upstream? Let's see if rewriting is necessary.

File: vllm_ascend/_310p/worker/v2/model_state.py [hunk 5]
```diff
@@ -13,34 +13,33 @@
...
+        self.rope_state = get_310p_rope_state(
+            self.model_config,
+            self.model,
+            self.max_num_reqs,
+            self.max_num_tokens,
+            self.max_model_len,
+            self.device,
+        )
+        try:
+            from vllm.v1.worker.gpu.model_states.mm_pruning import maybe_create_mm_pruner
```
Comment (YangShuai52) [inline vllm_ascend/_310p/worker/v2/model_state.py]: Please confirm if the 310PMRV1 supports EVS. If it doesn't, then the MRV2 probably won't support it either.

File: vllm_ascend/_310p/worker/v2/rope.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        prefill_token_ids: list[int],
+        mm_features: list,
+    ) -> None:
+        if self.has_delta:
+            mrope_model = cast(SupportsMRoPE, model)
+            prefill_positions, delta = mrope_model.get_mrope_input_positions(prefill_token_ids, mm_features)
+            self.prefill_delta.np[req_idx] = delta
+        else:
+            xdrope_model = cast(SupportsXDRoPE, model)
+            prefill_positions = xdrope_model.get_xdrope_input_positions(prefill_token_ids, mm_features)
```
Comment (YangShuai52) [inline vllm_ascend/_310p/worker/v2/rope.py]: Currently, the Qwen3-vl and Qwen3.5 versions compatible with the 310P do not require xdrope.

File: vllm_ascend/_310p/worker/v2/model_runner.py (hunk 4 above)
Comment (Thiagor2002) [reply vllm_ascend/_310p/worker/v2/model_runner.py:312]: Upstream GPU / non-310P Ascend MRv2 does not have this CPU gather race condition; 310P MRv1 has sync at the same position.

**Comment by Thiagor2002** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- E2E

**Comment by Thiagor2002** [comment]
/rerun

[Bot]: rerun completed.

**Rerun (failed jobs only):**
- E2E

**Comment by Thiagor2002** [comment]
/rerun

[Bot]: rerun completed.

**Rerun (failed jobs only):**
- E2E

**Comment by Thiagor2002** [comment]
/rerun

[Bot]: rerun completed.

**Rerun (failed jobs only):**
- E2E

**Comment by Thiagor2002** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- E2E

**Comment by Thiagor2002** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- E2E

**Review by YangShuai52 — APPROVED** [review]

**Comment by Thiagor2002** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- E2E

**Review by Tflowers-0129 — APPROVED** [review]

File: vllm_ascend/_310p/quantization/methods/w8a8_dynamic.py (hunk 2 above)
Comment (Thiagor2002) [reply vllm_ascend/_310p/quantization/methods/w8a8_dynamic.py]: it has been accepted and done

File: vllm_ascend/_310p/worker/v2/model_state.py (hunk 5 above)
Comment (Thiagor2002) [reply vllm_ascend/_310p/worker/v2/model_state.py]: it has been accepted and done.

File: vllm_ascend/_310p/worker/v2/rope.py (hunk 6 above)
Comment (Thiagor2002) [reply vllm_ascend/_310p/worker/v2/rope.py]: it has been accepted and done.

File: vllm_ascend/_310p/worker/v2/rope.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+        for dim in range(self.num_dims):
+            self.prefill_positions.stage_write(
+                self.num_dims * req_idx + dim,
+                0,
+                prefill_positions[dim].tolist(),
+            )
```
[the remark this answers is not in the document]
Comment (Thiagor2002) [reply vllm_ascend/_310p/worker/v2/rope.py:57]: Thanks, but this is not a bug. `SupportsMRoPE.get_mrope_input_positions` is specified to return `tuple[torch.Tensor, int]` with shape `[num_dims, num_tokens]`. Indexing `prefill_positions[dim]` yields a 1D tensor, so `.tolist()` is correct. Treating it as a Python list would fight the upstream interface. No change needed.

File: vllm_ascend/_310p/worker/v2/rope.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    if model_config.uses_xdrope_dim > 0:
+        assert isinstance(model, SupportsXDRoPE)
+        return Ascend310PRopeState(
+            model_config.uses_xdrope_dim,
+            False,
+            max_num_reqs,
+            max_num_tokens,
+            max_model_len,
+            device,
+        )
```
[the remark this answers is not in the document]
Comment (Thiagor2002) [reply vllm_ascend/_310p/worker/v2/rope.py]: We will keep direct `model_config.uses_mrope` access. It is a defined `ModelConfig` property, not an optional attribute, and the same pattern is used across Ascend / vLLM runners. Wrapping it in `getattr` is unnecessary defensive coding. XD-RoPE is explicitly out of scope for this 310P MRoPE path, so adding `uses_xdrope_dim` getattr guards is also not needed here.

File: vllm_ascend/_310p/worker/v2/model_state.py [hunk 9]
```diff
@@ -84,6 +83,85 @@ def prepare_attn(
...
+        self.rope_state.prepare_positions_cpu(
+            input_batch.idx_mapping_np,
+            input_batch.query_start_loc_np,
+            req_states.prefill_len.np,
+            req_states.num_computed_tokens_np,
+            input_batch.num_tokens_after_padding,
+        )
+        positions = self.rope_state.get_positions(input_batch.num_tokens_after_padding)
+        if self.model_config.uses_mrope:
+            prepare_mrope_cos_sin_slices_from_runner(self, positions)
```
[the remark this answers is not in the document]
Comment (Thiagor2002) [reply vllm_ascend/_310p/worker/v2/model_state.py:117]: We will keep direct `model_config.uses_mrope` access. It is a defined `ModelConfig` property, not an optional attribute, and the same pattern is used across Ascend / vLLM runners. Wrapping it in `getattr` is unnecessary defensive coding. XD-RoPE is explicitly out of scope for this 310P MRoPE path, so adding `uses_xdrope_dim` getattr guards is also not needed here.
